### ISSUE TITLE
TIFF/Old Style JEPG: javax.imageio.IIOException: Invalid JPEG file structure: missing SOS marker

### ISSUE THREAD

**Issue by davidszeto** [haraldk/TwelveMonkeys#145]
Hi, I have multi-page tiff with a mix of CCITT GROUP 4 and Old Style JPEG format. I have no trouble parsing the tiff file until it stumbles at the Old Style JPEG page which throws an exception:

javax.imageio.IIOException: Invalid JPEG file structure: missing SOS marker

I have been looking on the web for resolution with no luck.

Are you able to share some light on this issue?

Thanks
David

**Comment by haraldk** [comment]
Hi David,

TIFF compression 6/Old Style JPEG is often problematic, as it isn't a well defined spec, rather a number of different and sometimes incompatible implementations. If you can avoid it, it really is better to use a different compression. I'm not able to say much more about the exception you get, without actually seeing the file in question. 

Are you using my `TIFFImageReader` to read the TIFF (the reason I ask is it doesn't yet read CCITT group 4, however it *tries* to read Old Style JPEG)? For now, it tries to be reasonable and decode the JPEG stream, reconstructing the JPEG stream from the various JPEG TIFF tags if needed. It could be that it just doesn't read the particular implementation used by your writer. If so, I could probably be able to fix it, with a couple of sample files. 

Or it could be that the JPEG stream in the file is really missing a SOS marker, in which case there's not much to do (unless we can re-create it correctly using other info from the TIFF structure).

So... If you can send me some samples I'll look into it!


Best regards,

-- 
Harald K

**Comment by davidszeto** [comment]
Hi Harald, Thanks for your quick response.

Yes I use your TiffImageReader. It has no issue reading pages with CCITT group 4 but stumbles at pages with Old style jpeg.

Is there another way to verify the file is missing a SOS marker?

I am trying to get hold of a sample file for you. I will be in touch again soon.

Regards
David

**Comment by haraldk** [comment]
To verify if the JPEG SOS segment is there, you need a sample file. If you know the TIFF/JPEG structures, you should be able to verify this quite easily using a hex viewer/editor. If you send me a sample, I'll tell you right away. :-)

My guess is that the JPEG SOS segment is there, but not at the offset where my `TIFFImageReader` expects it to be (which again could be the writer software's or my fault, most likely mine...). 


Regards,

-- 
Harald K

**Comment by haraldk** [comment]
Hi David,

Have you found a sample file for me? :-)

-- 
Harald K

**Comment by davidszeto** [comment]
Hi Harald,

Sorry I am still waiting for the samples. I have some production files which have client info so I can't provide them to you.

David

**Comment by haraldk** [comment]
Hi David,

I'm closing this issue for now, as it's not much I can do about it without the failing samples. I've searched around, but unfortunately, "old style JPEG" TIFF files are hard to find. The few I have found, have all been read correctly by the library.

That said, I'm still very thankful if you (your client) can provide such sample files in the future! If/when that happens, don't hesitate to reopen the issue, attach the samples, and I will look into it.


Best regards,

-- 
Harald K
